# Determining Even/Odd Numbers
* Difficulty: Easy 
* Topics: Basic Programming
* Description: Write a program to check whether a number is even or odd.
* Example:
* Input: number = 4
* Output: Even
* Explanation: Since 4 is divisible by 2, it is an even number.

In [75]:
number = int(input("Enter a no"))
if number // 2 == 0:
    print(number, "is even")
else:
    print(number, "is odd")


22 is odd


# Checking for Prime Numbers
* Difficulty: Easy
* Topics: Basic Programming, Number Theory
* Description: Write a program to determine if a number is prime.
* Example:
* Input: number = 7
* Output: Prime
* Explanation: 7 has no divisors other than 1 and itself, so it is a prime number.

In [76]:
import math
number = int(input("Enter a number"))
is_prime = True

if number < 2:
    is_prime = False
else:
    for i in range(2, int(math.sqrt(number))+1):
        if number % i == 0:
            is_prime = False

if is_prime:
    print(number, "is a prime number")
else:
    print(number, "is not a prime number")

23 is a prime number


# Validating Leap Years
* Difficulty: Easy
* Topics: Basic Programming, Date Handling
* Description: Write a program to check if a given year is a leap year.
* Example:
* Input: year = 2020
* Output: Leap Year
* Explanation: 2020 is divisible by 4 but not by 100, or it is divisible by 400, so it is a leap year.

In [77]:
year = int(input("Enter the year"))

is_leapYear = True

if year % 4 == 0:
    if year % 100 == 0:
        if year % 400 == 0:
            is_leapYear = True

else:
    is_leapYear = False
    
if is_leapYear:
    print(f"{year} is a leap year")
else:
    print(f"{year} is not a leap year")

2024 is a leap year


# Calculating Armstrong Numbers
* Difficulty: Easy
* Topics: Basic Programming, Number Theory
* Description: Write a program to check if a number is an Armstrong number.
* Example:
* Input: number = 153
* Output: Armstrong Number
* Explanation: 153 is an Armstrong number because 1^3 + 5^3 + 3^3 = 153

In [78]:
number = int(input("Enter a number"))
is_armstrong = True

stringNumber = str(number)
x = len(stringNumber)
final = 0
for i in range(0,x):
    # print(stringNumber[i])
    final += int(stringNumber[i])**x
    
if final != number:
    is_armstrong = False
    
# print(final)
if is_armstrong:
    print(number,"is an Armstrong number")
else:
    print(number,"is not an Armstrong number")
    
    
    
    
    


123 is not an Armstrong number


In [79]:
"""Using maths only"""
number = int(input("Enter a number"))
original_number = number

# Step 1 ---> counting the no of digit
temp = number
count = 0  # counting the no of digit
while temp > 0:
    temp = temp // 10
    count += 1
    
# step 2 ---> exponential Summation of digits
sum_of_digits = 0
temp = number
while temp > 0:
    digit = temp % 10
    sum_of_digits += digit**count
    temp = temp // 10

if sum_of_digits == original_number:
    print(f"{original_number} is a armstrong number")
else:
    print(f"{original_number} is not a armstrong number")

153 is a armstrong number


# Generating the Fibonacci Series
* Difficulty: Easy
* Topics: Basic Programming, Sequences
* Description: Write a program to generate the Fibonacci series up to a given number.
* Example:
* Input: limit = 10
* Output: [0, 1, 1, 2, 3, 5, 8]
* Explanation: The Fibonacci series up to 10 is generated as [0, 1, 1, 2, 3, 5, 8].

In [80]:
length = int(input("Enter the length of the fibonacci series"))
fibList = [0,1]
if length == 0:
    print("0")

if length == 1:
    print(fibList)
else:
    while length > 1:
        fibList.append(fibList[-1] + fibList[-2])
        length -= 1
        
print(fibList)



[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


#   Identifying Palindromes
* Difficulty: Easy
* Topics: Basic Programming, String Manipulation
* Description: Write a program to check if a string or number is a palindrome.
* Example:
* Input: string = "radar"
* Output: Palindrome
* Explanation: "radar" reads the same backward as forward.

In [81]:
def is_palindrome(word:int) -> bool:
    """Check if a given integer is a palindrome."""
    if word == word[::-1]:
        return True
    else:
        return False

word = input("Enter the word")
if is_palindrome(word):
    print("The word is a palindrome")
else:
    print("The word is not a palindrome")
    

The word is a palindrome


# Crafting Star Patterns
* Difficulty: Easy
* Topics: Basic Programming, Patterns
* Description: Write a program to create different star patterns (e.g., pyramid, diamond).
* Example:
* Input: patternType = "pyramid", height = 5
* Output:


In [82]:
def pyramid(height):
    for i in range(1, height + 1):
        for j in range(2*height):
            if j >= (height+1)-i and j <= (height-1)+i:
                print("*",end="")
            else:
                print(" ",end="")
        print("")
height = int(input("Enter the height of the pyramid"))
pyramid(height)

    *   
   ***  
  ***** 
 *******


# Finding the Factorial of a Number
* Difficulty: Easy
* Topics: Basic Programming, Mathematical Computations
* Description: Write a program to compute the factorial of a given number.
* Example:
* Input: number = 5
* Output: 120
* Explanation: 5! (factorial) is 5 × 4 × 3 × 2 × 1 = 120.

In [3]:
# number = int(input("Enter the number"))
# for i in range (1,number):
#     number = number * i
    
# print(number)

def factorial(n):
    if n == 0:
        return 1
    if n == 1:
        return n
    else:
        return n * factorial(n-1)
    
n = int(input("Enter the number"))
print(factorial(n))
    

120


# Generating a Pattern of Numbers
* Difficulty: Easy
* Topics: Basic Programming, Patterns
* Description: Write a program to generate number patterns (e.g., sequential numbers in a matrix).
* Example:
* Input: rows = 3
* Output:<br>
1<br>
23<br>
456

In [4]:
count = 1
for i in range (1, 4):
    for j in range(1,4):
        if j<=i:
            print(count,end="")
            count += 1
        
    print("")

1
23
456


# Generating a Pascal’s Triangle
* Difficulty: Medium
* Topics: Arrays, Mathematical Computations
* Description: Write a program to generate Pascal's Triangle up to a given number of rows.
* Example:
* Input: rows = 4
* Output:

In [6]:
def factorial(n):
    if n == 0:
        return 1
    if n == 1:
        return n
    else:
        return n * factorial(n-1)

def combination(n:int, r:int)-> int:
    return (factorial(n)//(factorial(n-r)*factorial(r)))

def pascal_triangle(height:int)->None:
    for i in range (1,height+1):
        k = True
        r=0
        for j in range (2*height):
            if j >= (height + 1) - i and j <= (height -1) +i and k:
                print(combination(i-1,r), end="")
                k = False
                r += 1
            else:
                print(" ", end="")
                k = True
        print("")
        
height = int(input("enter the height"))
pascal_triangle(height)

     1    
    1 1   
   1 2 1  
  1 3 3 1 
 1 4 6 4 1


In [86]:
from math import factorial

def combination(n: int, r: int) -> int:
    return factorial(n) // (factorial(r) * factorial(n - r))

def pascal_triangle(height: int) -> None:
    # Determine the max width needed for the largest number
    max_num = combination(height - 1, (height - 1) // 2)
    cell_width = len(str(max_num)) + 2  # +2 for padding between numbers

    for i in range(height):
        line = ""
        for j in range(i + 1):
            num = str(combination(i, j)).center(cell_width)
            line += num
        # Center the line according to total width
        print(line.center(cell_width * height))

# Run
height = int(input("Enter the height of Pascal's Triangle: "))
pascal_triangle(height)

             1              
           1   1            
         1   2   1          
       1   3   3   1        
     1   4   6   4   1      
   1   5   10  10  5   1    
 1   6   15  20  15  6   1  


# Finding the Longest Substring Without Repeating Characters
* Difficulty: Medium
* Topics: String Manipulation, Sliding Window
* Description: Write a program to find the longest substring without repeating characters in a given string.
* Example:
* Input: string = "abcabcbb"
* Output: "abc"
* Explanation: The longest substring without repeating characters is "abc".

# Finding the Summation of Digits of a Number Until Zero
* Difficulty: Easy
* Topics: Basic Programming, Mathematical Computations
* Description: Write a program to repeatedly Summation the digits of a number until the result is zero.
* Example:
* Input: number = 123
* Output: 6
* Explanation: Summation of digits is 1 + 2 + 3 = 6; Summation of digits of 6 is 6 (which is a single digit).

In [89]:
num = int(input("Enter a number"))
while num >= 10:
    sum_digit = 0
    temp = num
    while temp > 0:
        sum_digit += temp % 10
        temp = temp // 10
    num = sum_digit
    
print(sum_digit) 

9


# Finding All Pairs of Elements in an Array that Add Up to a Given Summation
* Difficulty: Medium
* Topics: Arrays, Hashing
* Description: Write a program to find all pairs of elements in an array whose Summation equals a specified target.
* Example:
* Input: array = [1, 2, 3, 4, 5], target = 5
* Output: [(1, 4), (2, 3)]
* Explanation: Pairs that Summation up to 5 are (1, 4) and (2, 3).



In [90]:
# input_array = list(map(int,input("enter the number array").split()))
# target = int(input("Enter the target Summation: "))
# sol = []
# for i in range(len(input_array)):
#     for j in range(i+1, len(input_array)):
#         if input_array[i] + input_array[j] == target:
#             sol.append([input_array[i],input_array[j]])

# print(sol)

# two pointer
input_array = list(map(int, input("Enter the number array: ").split()))
target = int(input("Enter the target Summation: "))
input_array.sort()

left = 0
right = len(input_array) - 1
sol = []

while left < right:
    current_sum = input_array[left] + input_array[right]
    if current_sum == target:
        sol.append([input_array[left], input_array[right]])
        left += 1
        right -= 1
    elif current_sum < target:
        left += 1
    else:
        right -= 1

print(sol)


[[1, 5], [2, 4]]


# Generating a Diamond Pattern of Stars
* Difficulty: Medium
* Topics: Patterns, Basic Programming
* Description: Write a program to create a diamond pattern with stars of a given size.
* Example:
* Input: size = 5
* Output:



In [91]:
height = (int(input("Enter the height of the rectangle")))
for i in range(1,height+1):
    for j in range(2*height):
        if j>= (height - i) + 1 and j <= (height + i) - 1:
            print("*",end=" ")
        else:
            print(" ",end=" ")
    print("")
    
for i in range(height-1,0,-1):
    for j in range(2*height):
        if j>= (height - i) + 1 and j <= (height + i) - 1:
            print("*",end=" ")
        else:
            print(" ",end=" ")
    print("")


                *               
              * * *             
            * * * * *           
          * * * * * * *         
        * * * * * * * * *       
      * * * * * * * * * * *     
    * * * * * * * * * * * * *   
  * * * * * * * * * * * * * * * 
    * * * * * * * * * * * * *   
      * * * * * * * * * * *     
        * * * * * * * * *       
          * * * * * * *         
            * * * * *           
              * * *             
                *               


# Finding the GCD of Multiple Numbers
* Difficulty: Medium
* Topics: Mathematical Computations
* Description: Write a program to find the GCD (Greatest Common Divisor) of an array of numbers.
* Example:
* Input: array = [12, 24, 36]
* Output: 12
* Explanation: The GCD of 12, 24, and 36 is 12.

# Finding the Summation of Numbers in a String
* Difficulty: Easy
* Topics: String Manipulation
* Description: Write a program to extract and Summation all numbers present in a given string.
* Example:
* Input: string = "The numbers are 12 and 34"
* Output: 46
* Explanation: The Summation of numbers 12 and 34 is 46.

In [95]:
import math
string = "The numbers are 12 and 34"
result = [int(word) for word in string.split() if word.isdigit()]
print(math.fsum(result))


46.0


# Finding the Longest Consecutive Sequence in an Array
* Difficulty: Medium
* Topics: Arrays, Hashing
* Description: Write a program to find the longest sequence of consecutive numbers in an array.
* Example:
* Input: array = [100, 4, 200, 1, 3, 2]
* Output: 4
* Explanation: The longest consecutive sequence is [1, 2, 3, 4].

# Generating a Matrix with a Spiral Pattern
* Difficulty: Medium
* Topics: Arrays, Matrix Operations
* Description: Write a program to generate a matrix filled with numbers in a spiral pattern.
* Example:
* Input: size = 3
* Output:

# Generating Pascal’s Triangle Up to N Rows
* Difficulty: Medium
* Topics: Combinatorics
* Description: Write a program to generate Pascal’s Triangle up to N rows.
* Example:
* Input: N = 3
* Output:

# Generating Pascal’s Triangle Up to N Rows
* Difficulty: Medium
* Topics: Combinatorics
* Description: Write a program to generate Pascal’s Triangle up to N rows.
* Example:
* Input: N = 3
* Output:

1 0 0 0  
1 1 0 0  
1 1 1 0  
1 1 1 1

# Checking for Anagram Pairs in a List of Strings
* Difficulty: Medium
* Topics: String Manipulation
* Description: Write a program to find pairs of strings in a list that are anagrams of each other.
* Example:
* Input: strings = ["listen", "silent", "hello", "world"]
* Output: [("listen", "silent")]
* Explanation: "listen" and "silent" are anagrams.

# Finding All Triplets in an Array That Summation to Zero
* Difficulty: Medium
* Topics: Arrays, Sorting
* Description: Write a program to find all unique triplets in an array that Summation to zero.
* Example:
* Input: array = [-1, 0, 1, 2, -1, -4]
* Output: [[-1, -1, 2], [-1, 0, 1]]

# Problem 5: Print a Hollow Square of Stars
* Difficulty: Easy
* Topics: Pattern Printing
* Hint: Print a hollow square pattern with stars (*). The border of the square should be filled with stars while the inner part should be empty.
* Example 1: Input: n = 5
* Output:

```
*****
*   *
*   *
*   *
*****
```

# Problem 19: Print an Inverted Pyramid Pattern with Characters
* Difficulty: Medium
* Topics: Pattern Printing
* Hint: Print an inverted pyramid pattern using characters. Each row should have decreasing characters from the top row.
* 
* Example 1: Input: n = 3
* Output:

```
CCC
 BB
  A
  ```

In [ ]:
height = int(input("Enter the height"))
for i in range(height):
    for j in range(height):
        if j>=i:
            print("*", end=" ")
        else:
            print(" ",end="")
    print("")

* * * * 
 * * * 
  * * 
   * 
    
    
   * 
  * * 


# Problem 28: Print a Pattern with Diagonal Lines
* Difficulty: Hard
* Topics: Pattern Printing
* Hint: Print a pattern with diagonal lines using characters. Each diagonal line should be aligned properly.
* 
* Example 1: Input: n = 4
* Output:

```
A
B B
C   C
D     D

In [66]:
arr = ['A', 'B', 'C', 'D']

for i in range(1, 5):  # 4 rows
    for j in range(i):
        if j == 0 or j == i - 1:
            print(arr[i - 1], end='')
        else:
            print(' ', end='')
    print()

A
BB
C C
D  D


In [68]:
height = int(input("Enter height"))
for i in range(height):  # 4 rows
    ch = chr(65 + i)  # Convert number to corresponding uppercase letter
    for j in range(i + 1):
        if j == 0 or j == i:
            print(ch, end='')
        else:
            print(' ', end='')
    print()

A
BB
C C
D  D
E   E
F    F


# Problem 29: Print a Matrix with Diamond Pattern of Numbers
* Difficulty: Hard
* Topics: Matrix Pattern
* Hint: Print a matrix where elements follow a diamond pattern with numbers.
* 
* Example 1: Input: n = 3
* Output:

 ```
  1
 121
12321
 121
  1

In [100]:
height = (int(input("Enter the height of the rectangle")))
for i in range(1,height+1):
    count =1
    # ----------> Incrementing <-----------
    for j in range(height+1):
        if j>= (height - i) + 1:
            print(count,end=" ")
            count +=1
        else:
            print(" ",end=" ")
    # -----> decrementing <---------
    count -= 2    
    for j in range(1,height):
        if j<i:
            print(count,end=" ")
            count +=1
        else:
            print(" ",end=" ")
    print("")
    
# ---------->>> Lower diamond <<<-----------

 # ----------> Incrementing <-----------
for i in range(height-1,0,-1):
    count =1
    for j in range(height+1):
        if j>= (height - i) + 1:
            print(count,end=" ")
            count +=1
        else:
            print(" ",end=" ")
    # -----> decrementing <---------
    count -= 2    
    for j in range(1,height):
        if j<i:
            print(count,end=" ")
            count +=1
        else:
            print(" ",end=" ")
    print("")

      1     
    1 2 1   
  1 2 3 2 3 
    1 2 1   
      1     


# Problem 30: Print a Cross Pattern of Stars with Diagonals
* Difficulty: Hard
* Topics: Pattern Printing
* Hint: Print a cross pattern using stars with intersecting diagonals.
* 
* Example 1: Input: n = 5
* Output:

```
* * * * *
 * * * *
  * * *
   * *
    *
   * *
  * * *
 * * * *
* * * * *

In [32]:
height = int(input("enter the height"))

for i in range (height,0,-1):
    for j in range (2*height):
        if j >= height-i+1 and j <= height+i-1:
            print("*",end="")
        else:
            print(" ",end="")
    print("")
    
for i in range (2,height+1):
    for j in range (2*height):
        if j >= height-i+1 and j <= height+i-1:
            print("*",end="")
        else:
            print(" ",end="")
    print("")
            

 *********
  ******* 
   *****  
    ***   
     *    
    ***   
   *****  
  ******* 
 *********
